File: Sources/RefId/Canonical.swift
// SPDX-License-Identifier: Apache-2.0
//
// The canonical serialisation `spec.canonicalisation.rules` fix: object keys sorted by UTF-16 code
// unit, no whitespace outside strings, strings escaped exactly as ECMAScript JSON.stringify does,
// integers only. Written for the object graph Foundation's JSONSerialization produces.

import Foundation

enum Canonical {
    static func serialise(_ value: Any) throws -> String {
        var out = ""
        try write(value, into: &out)
        return out
    }

    private static func utf16Less(_ a: String, _ b: String) -> Bool {
        a.utf16.lexicographicallyPrecedes(b.utf16)
    }

    private static func write(_ value: Any, into out: inout String) throws {
        switch value {
        case is NSNull:
            out += "null"
        case let number as NSNumber:
            if CFGetTypeID(number) == CFBooleanGetTypeID() {
                out += number.boolValue ? "true" : "false"
            } else if let integer = integerValue(number) {
                out += String(integer)
            } else {
                throw RefIdError.specIntegrity("canonicalisation covers integers only, got \(number)")
            }
        case let string as String:
            out += escape(string)
        case let array as [Any]:
            out += "["
            for (index, item) in array.enumerated() {
                if index > 0 { out += "," }
                try write(item, into: &out)
            }
            out += "]"
        case let dictionary as [String: Any]:
            out += "{"
            for (index, key) in dictionary.keys.sorted(by: utf16Less).enumerated() {
                if index > 0 { out += "," }
                out += escape(key)
                out += ":"
                try write(dictionary[key]!, into: &out)
            }
            out += "}"
        default:
            throw RefIdError.specIntegrity("value of type \(Swift.type(of: value)) has no canonical form")
        }
    }

    /// An NSNumber that JSONSerialization parsed from an integer literal; a Double is refused.
    private static func integerValue(_ number: NSNumber) -> Int64? {
        let objCType = String(cString: number.objCType)
        switch objCType {
        case "d", "f":
            let double = number.doubleValue
            return double == double.rounded() && abs(double) < 9_007_199_254_740_992 ? Int64(double) : nil
        default:
            return number.int64Value
        }
    }

    /// ECMAScript JSON.stringify's string escaping: quote, backslash and control characters only.
    static func escape(_ string: String) -> String {
        var out = "\""
        for scalar in string.unicodeScalars {
            switch scalar {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\u{08}": out += "\\b"
            case "\u{0C}": out += "\\f"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            default:
                if scalar.value < 0x20 {
                    out += String(format: "\\u%04x", scalar.value)
                } else {
                    out.unicodeScalars.append(scalar)
                }
            }
        }
        out += "\""
        return out
    }
}

File: crates/ref-id/src/canonical.rs
// SPDX-License-Identifier: Apache-2.0
//
// The canonical serialisation `spec.canonicalisation.rules` fix: object keys sorted by UTF-16 code
// unit, no whitespace outside strings, strings escaped exactly as ECMAScript JSON.stringify does,
// integers only.

use crate::types::RefIdError;
use serde_json::Value;

/// ECMAScript JSON.stringify's string escaping: quote, backslash and control characters only.
pub(crate) fn escape(string: &str, out: &mut String) {
    out.push('"');
    for c in string.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0C}' => out.push_str("\\f"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
}

fn write(value: &Value, out: &mut String) -> Result<(), RefIdError> {
    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                out.push_str(&i.to_string());
            } else if let Some(u) = n.as_u64() {
                out.push_str(&u.to_string());
            } else {
                return Err(RefIdError::SpecIntegrity(format!("canonicalisation covers integers only, got {n}")));
            }
        }
        Value::String(s) => escape(s, out),
        Value::Array(items) => {
            out.push('[');
            for (i, item) in items.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                write(item, out)?;
            }
            out.push(']');
        }
        Value::Object(map) => {
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort_by(|a, b| a.encode_utf16().cmp(b.encode_utf16()));
            out.push('{');
            for (i, key) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                escape(key, out);
                out.push(':');
                write(&map[*key], out)?;
            }
            out.push('}');
        }
    }
    Ok(())
}

/// The canonical serialisation this crate computes the specification digest over.
pub fn canonicalise(value: &Value) -> Result<String, RefIdError> {
    let mut out = String::new();
    write(value, &mut out)?;
    Ok(out)
}
